File: langs/clojure/adapter.clj
;; argv: clojure -M adapter.clj <src> <class> <cases.json>
;; Load the solution, instantiate Simulation or InMemoryDatabase,
;; call snake_case methods, compare JSON encodings.
;; Bank booleans are JSON true/false.

(ns honepad.adapter)

(defn- make-parser [text]
  (atom {:text text :i 0 :n (count text)}))

(defn- peek-ch [p]
  (let [{:keys [text i n]} @p]
    (when (< i n)
      (.charAt ^String text i))))

(defn- bump! [p]
  (let [c (peek-ch p)]
    (when c
      (swap! p update :i inc))
    c))

(defn- skipws! [p]
  (loop []
    (let [c (peek-ch p)]
      (when (and c (#{\space \tab \newline \return} c))
        (bump! p)
        (recur)))))

(defn- starts-with? [p token]
  (let [{:keys [text i n]} @p
        m (count token)]
    (and (<= (+ i m) n)
         (= (.substring ^String text i (+ i m)) token))))

(defn- advance! [p nsteps]
  (swap! p update :i + nsteps))

(declare parse-value)

(defn- parse-string [p]
  (when-not (= (bump! p) \")
    (throw (ex-info "expected string" {})))
  (let [out (StringBuilder.)]
    (loop []
      (let [c (bump! p)]
        (cond
          (nil? c) (throw (ex-info "unterminated string" {}))
          (= c \") (str out)
          (= c \\)
          (let [e (bump! p)]
            (cond
              (#{\" \\ \/} e) (.append out e)
              (= e \b) (.append out \backspace)
              (= e \f) (.append out \formfeed)
              (= e \n) (.append out \newline)
              (= e \r) (.append out \return)
              (= e \t) (.append out \tab)
              (= e \u)
              (let [hex (apply str (repeatedly 4 #(or (bump! p)
                                                      (throw (ex-info "bad unicode" {})))))]
                (.append out (char (Integer/parseInt hex 16))))
              :else (throw (ex-info "bad escape" {})))
            (recur))
          :else
          (do (.append out c) (recur)))))))

(defn- parse-number [p]
  (let [start (:i @p)]
    (when (= (peek-ch p) \-)
      (bump! p))
    (while (let [c (peek-ch p)] (and c (Character/isDigit ^char c)))
      (bump! p))
    (let [float? (atom false)]
      (when (= (peek-ch p) \.)
        (reset! float? true)
        (bump! p)
        (while (let [c (peek-ch p)] (and c (Character/isDigit ^char c)))
          (bump! p)))
      (when-let [c (peek-ch p)]
        (when (#{\e \E} c)
          (reset! float? true)
          (bump! p)
          (when-let [s (peek-ch p)]
            (when (#{\+ \-} s)
              (bump! p)))
          (while (let [d (peek-ch p)] (and d (Character/isDigit ^char d)))
            (bump! p))))
      (let [text (.substring ^String (:text @p) start (:i @p))]
        (if @float?
          (let [n (Double/parseDouble text)
                i (Math/round n)]
            (if (= n (double i)) i n))
          (Long/parseLong text))))))

(defn- parse-array [p]
  (bump! p)
  (skipws! p)
  (if (= (peek-ch p) \])
    (do (bump! p) [])
    (loop [acc []]
      (let [acc (conj acc (parse-value p))]
        (skipws! p)
        (let [c (bump! p)]
          (cond
            (= c \]) acc
            (= c \,) (do (skipws! p) (recur acc))
            :else (throw (ex-info "expected comma or ]" {}))))))))

(defn- parse-object [p]
  (bump! p)
  (skipws! p)
  (if (= (peek-ch p) \})
    (do (bump! p) {})
    (loop [acc {}]
      (skipws! p)
      (let [k (parse-string p)]
        (skipws! p)
        (when-not (= (bump! p) \:)
          (throw (ex-info "expected colon" {})))
        (skipws! p)
        (let [acc (assoc acc k (parse-value p))]
          (skipws! p)
          (let [c (bump! p)]
            (cond
              (= c \}) acc
              (= c \,) (do (skipws! p) (recur acc))
              :else (throw (ex-info "expected comma or }" {})))))))))

(defn- parse-value [p]
  (skipws! p)
  (let [c (peek-ch p)]
    (cond
      (nil? c) (throw (ex-info "unexpected end" {}))
      (starts-with? p "null") (do (advance! p 4) nil)
      (starts-with? p "true") (do (advance! p 4) true)
      (starts-with? p "false") (do (advance! p 5) false)
      (= c \") (parse-string p)
      (= c \[) (parse-array p)
      (= c \{) (parse-object p)
      (or (= c \-) (and c (Character/isDigit ^char c))) (parse-number p)
      :else (throw (ex-info (str "unexpected json char " c) {})))))

(defn json-decode [text]
  (let [p (make-parser text)
        val (parse-value p)]
    (skipws! p)
    (when (peek-ch p)
      (throw (ex-info "trailing json" {})))
    val))
(defn- write-json-string [^String s ^StringBuilder out]
  (.append out \")
  (doseq [c s]
    (case c
      \" (.append out "\\\"")
      \\ (.append out "\\\\")
      \backspace (.append out "\\b")
      \formfeed (.append out "\\f")
      \newline (.append out "\\n")
      \return (.append out "\\r")
      \tab (.append out "\\t")
      (if (< (int c) 32)
        (.append out (format "\\u%04x" (int c)))
        (.append out c))))
  (.append out \"))

(defn- write-json [val ^StringBuilder out]
  (cond
    (false? val) (.append out "false")
    (true? val) (.append out "true")
    (nil? val) (.append out "null")
    (integer? val) (.append out (str val))
    (float? val)
    (let [i (Math/round (double val))]
      (if (= (double val) (double i))
        (.append out (str i))
        (.append out (str val))))
    (string? val) (write-json-string val out)
    (map? val)
    (do
      (.append out \{)
      (doseq [[i [k v]] (map-indexed vector val)]
        (when (pos? i)
          (.append out \,))
        (write-json (if (string? k) k (str k)) out)
        (.append out \:)
        (write-json v out))
      (.append out \}))
    (or (sequential? val) (vector? val))
    (do
      (.append out \[)
      (doseq [[i item] (map-indexed vector val)]
        (when (pos? i)
          (.append out \,))
        (write-json item out))
      (.append out \]))
    :else (throw (ex-info (str "cannot encode " (pr-str val)) {}))))

(defn json-encode [val]
  (let [out (StringBuilder.)]
    (write-json val out)
    (str out)))

(defn- coerce [val]
  (cond
    (float? val)
    (let [i (Math/round (double val))]
      (if (= (double val) (double i)) i val))
    (vector? val) (mapv coerce val)
    (sequential? val) (mapv coerce val)
    :else val))

(defn- fail-row [case-id index method expected actual]
  {"case" case-id
   "index" index
   "method" method
   "expected" expected
   "actual" actual})

(defn- load-solution [src]
  (let [n (create-ns 'honepad.user)]
    (binding [*ns* n]
      (clojure.core/refer-clojure)
      (load-file src))
    n))

(defn- new-target [user-ns class-name]
  (let [sym (symbol class-name)
        v (ns-resolve user-ns sym)]
    (when-not v
      (throw (ex-info (str "missing " class-name) {})))
    (v)))

(defn- invoke-method [user-ns obj method args]
  (let [v (ns-resolve user-ns (symbol method))]
    (when-not v
      (throw (ex-info "missing" {})))
    (apply v obj args)))

(defn- exc-name [err]
  (.getSimpleName (class err)))

(defn- call-result [user-ns obj case-id i call]
  (let [method (str (get call "m"))
        args (mapv coerce (or (get call "a") []))
        expected (coerce (get call "e"))]
    (try
      (let [actual (invoke-method user-ns obj method args)]
        (if (not= (json-encode actual) (json-encode expected))
          [:fail (fail-row case-id i method expected actual)]
          :ok))
      (catch Throwable err
        [:fail (fail-row case-id i method expected
                         (str "exc:" (exc-name err)))]))))

(defn- replay [user-ns class-name cases]
  (loop [rows cases
         passed 0
         failed []]
    (if (empty? rows)
      [passed failed]
      (let [row (first rows)
            obj (new-target user-ns class-name)
            case-id (str (get row "id"))
            calls (or (get row "calls") [])
            outcome
            (reduce (fn [acc [i call]]
                      (let [result (call-result user-ns obj case-id i call)]
                        (if (= result :ok)
                          acc
                          (reduced (conj acc (second result))))))
                    failed
                    (map-indexed vector calls))]
        (if (identical? outcome failed)
          (recur (rest rows) (inc passed) failed)
          (recur (rest rows) passed outcome))))))

(defn -main [& args]
  (when (not= (count args) 3)
    (binding [*out* *err*]
      (println "usage: clojure -M adapter.clj <src> <class> <cases.json>"))
    (System/exit 2))
  (let [[src class-name cases-path] args]
    (try
      (let [user-ns (load-solution src)]
        (when-not (ns-resolve user-ns (symbol class-name))
          (binding [*out* *err*]
            (println "missing class" class-name))
          (System/exit 2))
        (let [cases (json-decode (slurp cases-path))]
          (when-not (sequential? cases)
            (throw (ex-info "cases must be a json array" {})))
          (let [[passed failed] (replay user-ns class-name cases)
                payload {"passed" passed "failed" failed}]
            (println (json-encode payload))
            (System/exit (if (empty? failed) 0 1)))))
      (catch Throwable err
        (binding [*out* *err*]
          (println (.getMessage err)))
        (println (json-encode
                  {"passed" 0
                   "failed" [(fail-row "load" 0 "load" nil
                                       (str "exc:" (exc-name err)))]}))
        (System/exit 2)))))

(apply -main *command-line-args*)

File: langs/julia/adapter.jl
#!/usr/bin/env julia
# argv: julia adapter.jl <src> <class> <cases.json>
# include the solution, instantiate Simulation or InMemoryDatabase,
# call snake_case methods, compare JSON encodings.

module MiniJson
export decode, encode

mutable struct Parser
    s::String
    i::Int
    n::Int
end

function peek(p::Parser)
    return p.i > p.n ? UInt8(0) : codeunit(p.s, p.i)
end

function bump!(p::Parser)
    c = peek(p)
    p.i += 1
    return c
end

function skipws!(p::Parser)
    while p.i <= p.n
        c = codeunit(p.s, p.i)
        if c == 0x20 || c == 0x09 || c == 0x0a || c == 0x0d
            p.i += 1
        else
            break
        end
    end
end

function parse_string!(p::Parser)
    bump!(p) == 0x22 || error("expected string")
    buf = IOBuffer()
    while true
        p.i > p.n && error("unterminated string")
        c = bump!(p)
        if c == 0x22
            return String(take!(buf))
        elseif c == 0x5c
            e = bump!(p)
            if e == 0x22 || e == 0x5c || e == 0x2f
                write(buf, e)
            elseif e == 0x62
                write(buf, 0x08)
            elseif e == 0x66
                write(buf, 0x0c)
            elseif e == 0x6e
                write(buf, 0x0a)
            elseif e == 0x72
                write(buf, 0x0d)
            elseif e == 0x74
                write(buf, 0x09)
            elseif e == 0x75
                hex = String([Char(bump!(p)) for _ in 1:4])
                write(buf, Char(parse(UInt32, hex; base = 16)))
            else
                error("bad escape")
            end
        else
            write(buf, c)
        end
    end
end

function parse_number!(p::Parser)
    start = p.i
    peek(p) == 0x2d && bump!(p)
    while p.i <= p.n && 0x30 <= codeunit(p.s, p.i) <= 0x39
        bump!(p)
    end
    is_float = false
    if peek(p) == 0x2e
        is_float = true
        bump!(p)
        while p.i <= p.n && 0x30 <= codeunit(p.s, p.i) <= 0x39
            bump!(p)
        end
    end
    c = peek(p)
    if c == 0x65 || c == 0x45
        is_float = true
        bump!(p)
        c = peek(p)
        if c == 0x2b || c == 0x2d
            bump!(p)
        end
        while p.i <= p.n && 0x30 <= codeunit(p.s, p.i) <= 0x39
            bump!(p)
        end
    end
    text = p.s[start:(p.i - 1)]
    return is_float ? parse(Float64, text) : parse(Int, text)
end

function parse_array!(p::Parser)
    bump!(p)
    skipws!(p)
    acc = Any[]
    if peek(p) == 0x5d
        bump!(p)
        return acc
    end
    while true
        push!(acc, parse_value!(p))
        skipws!(p)
        c = bump!(p)
        if c == 0x5d
            return acc
        elseif c != 0x2c
            error("expected comma or ]")
        end
        skipws!(p)
    end
end

function parse_object!(p::Parser)
    bump!(p)
    skipws!(p)
    acc = Dict{String,Any}()
    if peek(p) == 0x7d
        bump!(p)
        return acc
    end
    while true
        skipws!(p)
        key = parse_string!(p)
        skipws!(p)
        bump!(p) == 0x3a || error("expected colon")
        skipws!(p)
        acc[key] = parse_value!(p)
        skipws!(p)
        c = bump!(p)
        if c == 0x7d
            return acc
        elseif c != 0x2c
            error("expected comma or }")
        end
    end
end

function parse_value!(p::Parser)
    skipws!(p)
    c = peek(p)
    if c == 0x6e
        p.s[p.i:(p.i + 3)] == "null" || error("expected null")
        p.i += 4
        return nothing
    elseif c == 0x74
        p.s[p.i:(p.i + 3)] == "true" || error("expected true")
        p.i += 4
        return true
    elseif c == 0x66
        p.s[p.i:(p.i + 4)] == "false" || error("expected false")
        p.i += 5
        return false
    elseif c == 0x22
        return parse_string!(p)
    elseif c == 0x5b
        return parse_array!(p)
    elseif c == 0x7b
        return parse_object!(p)
    elseif c == 0x2d || (0x30 <= c <= 0x39)
        return parse_number!(p)
    end
    error("unexpected json byte $c")
end

function decode(text::AbstractString)
    p = Parser(String(text), 1, ncodeunits(String(text)))
    val = parse_value!(p)
    skipws!(p)
    p.i > p.n || error("trailing json")
    return val
end

function write_val(io::IO, val::Nothing)
    print(io, "null")
end

function write_val(io::IO, val::Bool)
    print(io, val ? "true" : "false")
end

function write_val(io::IO, val::Integer)
    print(io, val)
end

function write_val(io::IO, val::AbstractFloat)
    if isinteger(val) && typemin(Int) <= val <= typemax(Int)
        print(io, Int(val))
    else
        print(io, val)
    end
end

function write_val(io::IO, val::AbstractString)
    print(io, '"')
    for c in val
        if c == '"'
            print(io, "\\\"")
        elseif c == '\\'
            print(io, "\\\\")
        elseif c == '\b'
            print(io, "\\b")
        elseif c == '\f'
            print(io, "\\f")
        elseif c == '\n'
            print(io, "\\n")
        elseif c == '\r'
            print(io, "\\r")
        elseif c == '\t'
            print(io, "\\t")
        elseif UInt32(c) < 0x20
            print(io, "\\u", lpad(string(UInt32(c); base = 16), 4, '0'))
        else
            print(io, c)
        end
    end
    print(io, '"')
end

function write_val(io::IO, val::AbstractVector)
    print(io, '[')
    for (i, item) in enumerate(val)
        i > 1 && print(io, ',')
        write_val(io, item)
    end
    print(io, ']')
end

function write_val(io::IO, val::Tuple)
    write_val(io, collect(val))
end

function write_val(io::IO, val::AbstractDict)
    print(io, '{')
    first = true
    for key in sort!(collect(keys(val)); by = string)
        first || print(io, ',')
        first = false
        write_val(io, string(key))
        print(io, ':')
        write_val(io, val[key])
    end
    print(io, '}')
end

function encode(val)
    buf = IOBuffer()
    write_val(buf, val)
    return String(take!(buf))
end

end

function fail_row(case_id, index, method, expected, actual)
    return Dict{String,Any}(
        "case" => case_id,
        "index" => index,
        "method" => method,
        "expected" => expected,
        "actual" => actual,
    )
end

function coerce(val)
    if val isa AbstractFloat
        i = round(Int, val)
        return val == Float64(i) ? i : val
    elseif val isa AbstractVector
        return Any[coerce(x) for x in val]
    else
        return val
    end
end

function exc_name(err)
    T = typeof(err)
    return string(nameof(T))
end

function new_target(class_name)
    T = getfield(Main, Symbol(class_name))
    return T()
end

function invoke_method(obj, method, args)
    fn = getfield(Main, Symbol(method))
    return fn(obj, args...)
end

function main()
    if length(ARGS) < 3
        println(stderr, "usage: julia adapter.jl <src> <class> <cases.json>")
        exit(2)
    end
    src, class_name, cases_path = ARGS[1], ARGS[2], ARGS[3]
    include(abspath(src))
    if !isdefined(Main, Symbol(class_name))
        println(stderr, "missing type $class_name")
        exit(2)
    end
    cases = MiniJson.decode(read(cases_path, String))
    failed = Any[]
    passed = 0
    for row in cases
        # include() defines types after this file compiled.
        obj = Base.invokelatest(new_target, class_name)
        case_id = string(row["id"])
        calls = Base.get(row, "calls", Any[])
        ok = true
        for i in eachindex(calls)
            call = calls[i]
            method = string(call["m"])
            args = Any[coerce(a) for a in Base.get(call, "a", Any[])]
            expected = coerce(call["e"])
            local actual
            try
                actual = Base.invokelatest(invoke_method, obj, method, args)
            catch err
                push!(
                    failed,
                    fail_row(case_id, i - 1, method, expected, "exc:$(exc_name(err))"),
                )
                ok = false
                break
            end
            if MiniJson.encode(actual) != MiniJson.encode(expected)
                push!(failed, fail_row(case_id, i - 1, method, expected, actual))
                ok = false
                break
            end
        end
        if ok
            passed += 1
        end
    end
    println(MiniJson.encode(Dict{String,Any}("failed" => failed, "passed" => passed)))
    exit(isempty(failed) ? 0 : 1)
end

main()
